File: Advent2022/Day3/Day3.py
from typing import Generator

import aoc_utils as au


def inboth(rucksack: str) -> str:
    return (
        set(rucksack[: len(rucksack) // 2]) & set(rucksack[len(rucksack) // 2 :])
    ).pop()


def score(char: str) -> int:
    return 1 + (ord(char) - ord("a") if char.islower() else 26 + ord(char) - ord("A"))


# expected 157
def test_one(input: list[str]) -> int:
    return sum([score(inboth(rucksack)) for rucksack in input])


# expected 7845
def part_one(input: list[str]) -> int:
    return sum([score(inboth(rucksack)) for rucksack in input])


def in_three(first: str, second: str, third: str) -> str:
    return (set(first) & set(second) & set(third)).pop()


# expected 70
def test_two(input: list[str]) -> int:
    return sum(
        score(in_three(*triplet))
        for triplet in zip(input[0::3], input[1::3], input[2::3])
    )


# expected 2790
def part_two(input: list[str]) -> int:
    return sum(
        score(in_three(*triplet))
        for triplet in zip(input[0::3], input[1::3], input[2::3])
    )


if __name__ == "__main__":
    day = au.Day(
        2022,
        3,
        test_one,
        test_two,
        part_one,
        part_two,
        input=au.Day.InType.INPUT_MULTI_LINE_STR,
        test_input=au.Day.InType.INPUT_MULTI_LINE_STR,
    )

    day.run_all(run_tests=True)

File: Advent2022/Day3/Day3.cpp
#include "aoc_utils.h"

#include <string>
#include <vector>
#include <set>
#include <algorithm>
#include <numeric>
#include <iterator>

#include <iostream>

using namespace std;

class Day3 : public au::Day<vector<string>>
{
private:
    string letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";

    vector<int> inboth(const vector<string> &input)
    {
        vector<int> priorities;

        for (string line : input)
        {
            set<wchar_t> left(line.begin(), line.begin() + line.size() / 2);
            set<wchar_t> right(line.begin() + line.size() / 2, line.end());

            set<wchar_t> both;
            set_intersection(left.begin(), left.end(), right.begin(), right.end(),
                             std::inserter(both, both.begin()));

            priorities.push_back(1 + letters.find(*both.begin()));
        }
        return priorities;
    }

    vector<int> in_three(const vector<string> &input)
    {
        vector<int> priorities;

        for (auto it = input.begin(); it != input.end(); advance(it, 3))
        {
            set<wchar_t> first(it->begin(), it->end());
            set<wchar_t> second((it + 1)->begin(), (it + 1)->end());
            set<wchar_t> third((it + 2)->begin(), (it + 2)->end());

            set<wchar_t> two;
            set_intersection(first.begin(), first.end(), second.begin(), second.end(),
                             inserter(two, two.begin()));
            set<wchar_t> three;
            set_intersection(two.begin(), two.end(), third.begin(), third.end(),
                             inserter(three, three.begin()));

            priorities.push_back(1 + letters.find(*three.begin()));
        }
        return priorities;
    }

public:
    Day3() : Day(2022, 3) {}

    int test_one(const vector<string> &input)
    {
        vector<int> priorities = inboth(input);

        return accumulate(priorities.begin(), priorities.end(), 0);
    }

    int part_one(const vector<string> &input)
    {
        vector<int> priorities = inboth(input);

        return accumulate(priorities.begin(), priorities.end(), 0);
    }

    int test_two(const vector<string> &input)
    {
        vector<int> priorities = in_three(input);

        return accumulate(priorities.begin(), priorities.end(), 0);
    }

    int part_two(const vector<string> &input)
    {
        vector<int> priorities = in_three(input);

        return accumulate(priorities.begin(), priorities.end(), 0);
    }
};

int main()
{
    Day3().run_all(true);
}
